### PR TITLE
implemented error message

### LINKED ISSUE #12624: Enable Hyperparameter logging from any hook in the LightningModule
## 🚀 Feature

Make it possible to call `save_hyperparameters` from any hook in the `LightningModule`.

### Motivation

Sometimes the dataset has hyperparameters that should be logged. However, the `LightningDataModule` is only accessible from the `LightningModule` once the trainer is initiated. Thus, it would be useful to call `save_hyperparameters` from `on_fit_start`, when the `Trainer` is specified and the hyperparameters from the dataset can easily be collected, e.g. through `self.trainer.datamodule.build_hparams()`.

### Pitch

`log_hyperparameters` shouldn't look for `init` args in the local variables when called outside the `__init__` method.

Currently, this behaviour casues an exception in line 154 in `utilities/parsing.py`
```
local_args = {k: local_vars[k] for k in init_parameters.keys()}
```
because the function is looking for the init parameters in the local variables, which are only available when called from `__init__`.

Suggestion: Remove init parameter logging when called from other places. 

### Alternatives

Save init parameters and add them later.

### Additional context

N/A

### PR DESCRIPTION
**Pull request by cemde**
## What does this PR do?

The LightningModule will now raise an exception if the save_hyperparameter method is called from outside the `__init__` method.

Fixes #12624

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/mixins/hparams_mixin.py [hunk 1]
```diff
@@ -47,6 +47,10 @@ class ``__init__`` to be ignored
             frame: a frame object. Default is None
             logger: Whether to send the hyperparameters to the logger. Default: True
 
+        Raises::
+            ValueError:
```
Comment (justusschock) [inline pytorch_lightning/core/mixins/hparams_mixin.py:51]: I guess this should be a `RuntimeError` or a `UserError` instead since it's not the value that's wrong.

File: pytorch_lightning/core/mixins/hparams_mixin.py [hunk 2]
```diff
@@ -47,6 +47,10 @@ class ``__init__`` to be ignored
             frame: a frame object. Default is None
             logger: Whether to send the hyperparameters to the logger. Default: True
 
+        Raises::
```
Comment (justusschock) [inline pytorch_lightning/core/mixins/hparams_mixin.py:50]:
```suggestion
        Raises:
```

File: pytorch_lightning/core/mixins/hparams_mixin.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/core/mixins/hparams_mixin.py:51]: I'd suggest [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/models/test_hparams.py) :)

**Comment by akihironitta** [comment]
Hi @cemde thank you for working on this. Could you address the above review comment? Let me know if you have any questions!

**Comment by cemde** [comment]
> 

Sorry been busy. Will add tests asap.

**Comment by cemde** [comment]
@akihironitta I added the changes.

My implementation breaks two tests:
- `tests/tests_pytorch/models/test_hparams.py::test_collect_init_arguments[UnconventionalArgsBoringModel]`
- `tests/tests_pytorch/models/test_hparams.py::test_dataclass_lightning_module`

Do these tests still reflect a wanted functionality? If so, the fix for the dataclass is easy. The `UnconventionalArgsBoringModel` does break the assumption of checkpointing that the init args are collected for this exact class to reconstruct it, doesnt it?

**Comment by cemde** [comment]
@akihironitta any opinion on this? :)
